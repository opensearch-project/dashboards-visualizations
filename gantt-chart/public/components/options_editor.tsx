/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiColorPicker,
  EuiFieldText,
  EuiFormRow,
  EuiPanel,
  EuiSelect,
  EuiSpacer,
  EuiSwitch,
  EuiTitle,
} from '@elastic/eui';
import React from 'react';
import {
  GanttParams,
  PlotlyAxisPosition,
  PlotlyAxisType,
  PlotlyLegendOrientation,
  VisOptionsProps,
} from '../gantt_vis_type';

export function OptionsEditor({ aggs, stateParams, setValue }: VisOptionsProps<GanttParams>) {
  const legendOrientationOptions: Array<{ value: PlotlyLegendOrientation; text: string }> = [
    { value: 'v', text: 'Vertical' },
    { value: 'h', text: 'Horizontal' },
  ];

  const renderLegendOptions = () => {
    return (
      <EuiPanel paddingSize="s">
        <EuiTitle size="xs">
          <h3>Legend</h3>
        </EuiTitle>
        <EuiSpacer size="s" />

        <EuiFormRow>
          <EuiSwitch
            data-test-subj="options-editor-legend-switch"
            label="Show legend"
            checked={stateParams.showLegend}
            onChange={(e) => setValue('showLegend', e.target.checked)}
          />
        </EuiFormRow>

        {stateParams.showLegend && (
          <EuiFormRow label="Orientation">
            <EuiSelect
              data-test-subj="options-editor-legend-orientation-select"
              options={legendOrientationOptions}
              value={stateParams.legendOrientation}
              onChange={(e) =>
                setValue('legendOrientation', e.target.value as PlotlyLegendOrientation)
              }
              disabled={!stateParams.showLegend}
            />
          </EuiFormRow>
        )}
      </EuiPanel>
    );
  };

  const renderGridOptions = () => {
    return (
      <EuiPanel paddingSize="s">
        <EuiTitle size="xs">
          <h3>Grid</h3>
        </EuiTitle>
        <EuiSpacer size="s" />

        <EuiFormRow>
          <EuiSwitch
            data-test-subj="options-editor-grid-y-switch"
            label="Show Y-axis grids"
            checked={stateParams.yAxisShowGrid}
            onChange={(e) => setValue('yAxisShowGrid', e.target.checked)}
          />
        </EuiFormRow>
        <EuiFormRow>
          <EuiSwitch
            data-test-subj="options-editor-grid-x-switch"
            label="Show X-axis grids"
            checked={stateParams.xAxisShowGrid}
            onChange={(e) => setValue('xAxisShowGrid', e.target.checked)}
          />
        </EuiFormRow>
      </EuiPanel>
    );
  };

  const renderColorOptions = () => {
    return (
      <EuiPanel paddingSize="s">
        <EuiTitle size="xs">
          <h3>Colors</h3>
        </EuiTitle>
        <EuiSpacer size="s" />

        <EuiFormRow label="Color">
          <EuiColorPicker color={stateParams.colors} onChange={(e) => setValue('colors', e)} />
        </EuiFormRow>
      </EuiPanel>
    );
  };

  const yAxisPositionOptions: Array<{ value: PlotlyAxisPosition; text: string }> = [
    { value: 'left', text: 'Left' },
    { value: 'right', text: 'Right' },
  ];

  const xAxisPositionOptions: Array<{ value: PlotlyAxisPosition; text: string }> = [
    { value: 'top', text: 'Top' },
    { value: 'bottom', text: 'Bottom' },
  ];

  const axisTypeOptions: Array<{ value: PlotlyAxisType; text: string }> = [
    { value: '-', text: 'Auto' },
    { value: 'linear', text: 'Linear' },
    { value: 'log', text: 'Log' },
  ];

  const timeFormatOptions: Array<{ value: string; text: string }> = [
    { value: 'hh:mm:ss.SSS A', text: 'hh:mm:ss.SSS (12 hours)' },
    { value: 'MM/DD hh:mm:ss A', text: 'MM/DD hh:mm:ss (12 hours)' },
    { value: 'MM/DD/YY hh:mm A', text: 'MM/DD/YY hh:mm (12 hours)' },
    { value: 'HH:mm:ss.SSS', text: 'hh:mm:ss.SSS (24 hours)' },
    { value: 'MM/DD HH:mm:ss', text: 'MM/DD hh:mm:ss (24 hours)' },
    { value: 'MM/DD/YY HH:mm', text: 'MM/DD/YY hh:mm (24 hours)' },
  ];

  const renderYAxisOptions = () => {
    return (
      <EuiPanel paddingSize="s">
        <EuiTitle size="xs">
          <h3>Y-axis</h3>
        </EuiTitle>
        <EuiSpacer size="s" />

        <EuiFormRow label="Position">
          <EuiSelect
            data-test-subj="options-editor-y-axis-position-select"
            options={yAxisPositionOptions}
            value={stateParams.yAxisPosition}
            onChange={(e) => setValue('yAxisPosition', e.target.value as PlotlyAxisPosition)}
          />
        </EuiFormRow>

        <EuiFormRow>
          <EuiSwitch
            data-test-subj="options-editor-y-axis-switch"
            label="Show Y-axis line"
            checked={stateParams.yAxisShowLine}
            onChange={(e) => setValue('yAxisShowLine', e.target.checked)}
          />
        </EuiFormRow>
        <EuiFormRow>
          <EuiSwitch
            data-test-subj="options-editor-y-axis-label-switch"
            label="Show Y-axis label"
            checked={stateParams.yAxisShowTitle}
            onChange={(e) => setValue('yAxisShowTitle', e.target.checked)}
          />
        </EuiFormRow>

        {stateParams.yAxisShowTitle ? (
          <EuiFormRow label="Label">
            <EuiFieldText
              data-test-subj="options-editor-y-axis-label-input"
              placeholder="Label"
              value={stateParams.yAxisTitle}
              onChange={(e) => setValue('yAxisTitle', e.target.value)}
            />
          </EuiFormRow>
        ) : null}
      </EuiPanel>
    );
  };

  const renderXAxisOptions = () => {
    return (
      <EuiPanel paddingSize="s">
        <EuiTitle size="xs">
          <h3>X-axis</h3>
        </EuiTitle>
        <EuiSpacer size="s" />

        <EuiFormRow label="Position">
          <EuiSelect
            data-test-subj="options-editor-x-axis-position-select"
            options={xAxisPositionOptions}
            value={stateParams.xAxisPosition}
            onChange={(e) => setValue('xAxisPosition', e.target.value as PlotlyAxisPosition)}
          />
        </EuiFormRow>
        <EuiFormRow label="Scale type">
          <EuiSelect
            data-test-subj="options-editor-x-axis-scale-type-select"
            options={axisTypeOptions}
            value={stateParams.xAxisType}
            onChange={(e) => setValue('xAxisType', e.target.value as PlotlyAxisType)}
          />
        </EuiFormRow>
        <EuiFormRow label="Time format">
          <EuiSelect
            data-test-subj="options-editor-x-axis-time-format-select"
            options={timeFormatOptions}
            value={stateParams.timeFormat}
            onChange={(e) => setValue('timeFormat', e.target.value)}
          />
        </EuiFormRow>

        <EuiFormRow>
          <EuiSwitch
            data-test-subj="options-editor-x-axis-switch"
            label="Show X-axis line"
            checked={stateParams.xAxisShowLine}
            onChange={(e) => setValue('xAxisShowLine', e.target.checked)}
          />
        </EuiFormRow>
        <EuiFormRow>
          <EuiSwitch
            data-test-subj="options-editor-x-axis-label-switch"
            label="Show X-axis label"
            checked={stateParams.xAxisShowTitle}
            onChange={(e) => setValue('xAxisShowTitle', e.target.checked)}
          />
        </EuiFormRow>

        {stateParams.xAxisShowTitle ? (
          <EuiFormRow label="Label">
            <EuiFieldText
              data-test-subj="options-editor-x-axis-label-input"
              placeholder="Label"
              value={stateParams.xAxisTitle}
              onChange={(e) => setValue('xAxisTitle', e.target.value)}
            />
          </EuiFormRow>
        ) : null}
      </EuiPanel>
    );
  };

  return (
    <>
      {renderYAxisOptions()}
      <EuiSpacer size="s" />
      {renderXAxisOptions()}
      <EuiSpacer size="s" />
      {renderLegendOptions()}
      <EuiSpacer size="s" />
      {renderGridOptions()}
      <EuiSpacer size="s" />
      {renderColorOptions()}
    </>
  );
}
