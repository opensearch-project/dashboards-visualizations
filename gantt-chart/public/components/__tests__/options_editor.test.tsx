/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { fireEvent, render } from '@testing-library/react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { mockEditorProps } from '../../../test/mocks/mockData';
import { OptionsEditor } from '../options_editor';

describe('<OptionsEditor /> spec', () => {
  configure({ adapter: new Adapter() });

  it('renders the component', () => {
    const setValue = jest.fn();
    const utils = render(<OptionsEditor {...mockEditorProps(setValue)} />);
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders the component with different options', () => {
    const setValue = jest.fn();
    const props = mockEditorProps(setValue);
    props.stateParams = {
      ...props.stateParams,
      showLegend: false,
      yAxisShowTitle: false,
      xAxisShowTitle: false,
    };
    const utils = render(<OptionsEditor {...props} />);
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('sets legends', () => {
    const setValue = jest.fn();
    const utils = render(<OptionsEditor {...mockEditorProps(setValue)} />);
    const orientation = utils.getByTestId('options-editor-legend-orientation-select');
    fireEvent.change(orientation, { target: { value: 'h' } });
    expect(setValue).toBeCalledWith('legendOrientation', 'h');
    utils.getByTestId('options-editor-legend-switch').click();
    expect(setValue).toBeCalledWith('showLegend', false);
  });

  it('toggles grid y axis lines', () => {
    const setValue = jest.fn();
    const utils = render(<OptionsEditor {...mockEditorProps(setValue)} />);
    utils.getByTestId('options-editor-grid-y-switch').click();
    expect(setValue).toBeCalledWith('yAxisShowGrid', true);
  });

  it('toggles grid x axis lines', () => {
    const setValue = jest.fn();
    const utils = render(<OptionsEditor {...mockEditorProps(setValue)} />);
    utils.getByTestId('options-editor-grid-x-switch').click();
    expect(setValue).toBeCalledWith('xAxisShowGrid', false);
  });

  it('sets colors', () => {
    const setValue = jest.fn();
    const utils = render(<OptionsEditor {...mockEditorProps(setValue)} />);
    const colorPicker = utils.getByTestId('colorPickerAnchor');
    fireEvent.change(colorPicker, { target: { value: '#EEEEEE' } });
    expect(setValue).toBeCalledWith('colors', '#EEEEEE');
  });

  it('sets y axis', () => {
    const setValue = jest.fn();
    const utils = render(<OptionsEditor {...mockEditorProps(setValue)} />);
    fireEvent.change(utils.getByTestId('options-editor-y-axis-position-select'), {
      target: { value: 'right' },
    });
    expect(setValue).toBeCalledWith('yAxisPosition', 'right');

    fireEvent.change(utils.getByTestId('options-editor-y-axis-label-input'), {
      target: { value: 'test-label' },
    });
    expect(setValue).toBeCalledWith('yAxisTitle', 'test-label');

    fireEvent.click(utils.getByTestId('options-editor-y-axis-switch'));
    expect(setValue).toBeCalledWith('yAxisShowLine', false);

    fireEvent.click(utils.getByTestId('options-editor-y-axis-label-switch'));
    expect(setValue).toBeCalledWith('yAxisShowTitle', false);
  });

  it('sets x axis', () => {
    const setValue = jest.fn();
    const utils = render(<OptionsEditor {...mockEditorProps(setValue)} />);
    fireEvent.change(utils.getByTestId('options-editor-x-axis-position-select'), {
      target: { value: 'top' },
    });
    expect(setValue).toBeCalledWith('xAxisPosition', 'top');

    fireEvent.change(utils.getByTestId('options-editor-x-axis-scale-type-select'), {
      target: { value: 'log' },
    });
    expect(setValue).toBeCalledWith('xAxisType', 'log');

    fireEvent.change(utils.getByTestId('options-editor-x-axis-time-format-select'), {
      target: { value: 'MM/DD hh:mm:ss A' },
    });
    expect(setValue).toBeCalledWith('timeFormat', 'MM/DD hh:mm:ss A');

    fireEvent.change(utils.getByTestId('options-editor-x-axis-label-input'), {
      target: { value: 'test-label' },
    });
    expect(setValue).toBeCalledWith('xAxisTitle', 'test-label');

    fireEvent.click(utils.getByTestId('options-editor-x-axis-switch'));
    expect(setValue).toBeCalledWith('xAxisShowLine', false);

    fireEvent.click(utils.getByTestId('options-editor-x-axis-label-switch'));
    expect(setValue).toBeCalledWith('xAxisShowTitle', false);
  });
});
