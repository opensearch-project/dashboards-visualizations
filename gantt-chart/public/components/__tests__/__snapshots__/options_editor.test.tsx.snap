// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OptionsEditor /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--paddingSmall"
>
  <h3
    class="euiTitle euiTitle--xsmall"
  >
    Y-axis
  </h3>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        Position
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiFormControlLayout"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <select
            class="euiSelect"
            data-test-subj="options-editor-y-axis-position-select"
            id="random_html_id"
          >
            <option
              value="left"
            >
              Left
            </option>
            <option
              value="right"
            >
              Right
            </option>
          </select>
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiSwitch"
      >
        <button
          aria-checked="true"
          aria-labelledby="random_html_id"
          class="euiSwitch__button"
          data-test-subj="options-editor-y-axis-switch"
          id="random_html_id"
          role="switch"
          type="button"
        >
          <span
            class="euiSwitch__body"
          >
            <span
              class="euiSwitch__thumb"
            />
            <span
              class="euiSwitch__track"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon-isLoading euiSwitch__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon-isLoading euiSwitch__icon euiSwitch__icon--checked"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
            </span>
          </span>
        </button>
        <span
          class="euiSwitch__label"
          id="random_html_id"
        >
          Show Y-axis line
        </span>
      </div>
    </div>
  </div>
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiSwitch"
      >
        <button
          aria-checked="true"
          aria-labelledby="random_html_id"
          class="euiSwitch__button"
          data-test-subj="options-editor-y-axis-label-switch"
          id="random_html_id"
          role="switch"
          type="button"
        >
          <span
            class="euiSwitch__body"
          >
            <span
              class="euiSwitch__thumb"
            />
            <span
              class="euiSwitch__track"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon-isLoading euiSwitch__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon-isLoading euiSwitch__icon euiSwitch__icon--checked"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
            </span>
          </span>
        </button>
        <span
          class="euiSwitch__label"
          id="random_html_id"
        >
          Show Y-axis label
        </span>
      </div>
    </div>
  </div>
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        Label
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiFormControlLayout"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <input
            class="euiFieldText"
            data-test-subj="options-editor-y-axis-label-input"
            id="random_html_id"
            placeholder="Label"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<OptionsEditor /> spec renders the component with different options 1`] = `
<div
  class="euiPanel euiPanel--paddingSmall"
>
  <h3
    class="euiTitle euiTitle--xsmall"
  >
    Y-axis
  </h3>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__labelWrapper"
    >
      <label
        class="euiFormLabel euiFormRow__label"
        for="random_html_id"
      >
        Position
      </label>
    </div>
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiFormControlLayout"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <select
            class="euiSelect"
            data-test-subj="options-editor-y-axis-position-select"
            id="random_html_id"
          >
            <option
              value="left"
            >
              Left
            </option>
            <option
              value="right"
            >
              Right
            </option>
          </select>
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                  fill-rule="non-zero"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiSwitch"
      >
        <button
          aria-checked="true"
          aria-labelledby="random_html_id"
          class="euiSwitch__button"
          data-test-subj="options-editor-y-axis-switch"
          id="random_html_id"
          role="switch"
          type="button"
        >
          <span
            class="euiSwitch__body"
          >
            <span
              class="euiSwitch__thumb"
            />
            <span
              class="euiSwitch__track"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiSwitch__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7.293 8L3.146 3.854a.5.5 0 11.708-.708L8 7.293l4.146-4.147a.5.5 0 01.708.708L8.707 8l4.147 4.146a.5.5 0 01-.708.708L8 8.707l-4.146 4.147a.5.5 0 01-.708-.708L7.293 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiSwitch__icon euiSwitch__icon--checked"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6.5 12a.502.502 0 01-.354-.146l-4-4a.502.502 0 01.708-.708L6.5 10.793l6.646-6.647a.502.502 0 01.708.708l-7 7A.502.502 0 016.5 12"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </span>
        </button>
        <span
          class="euiSwitch__label"
          id="random_html_id"
        >
          Show Y-axis line
        </span>
      </div>
    </div>
  </div>
  <div
    class="euiFormRow"
    id="random_html_id-row"
  >
    <div
      class="euiFormRow__fieldWrapper"
    >
      <div
        class="euiSwitch"
      >
        <button
          aria-checked="false"
          aria-labelledby="random_html_id"
          class="euiSwitch__button"
          data-test-subj="options-editor-y-axis-label-switch"
          id="random_html_id"
          role="switch"
          type="button"
        >
          <span
            class="euiSwitch__body"
          >
            <span
              class="euiSwitch__thumb"
            />
            <span
              class="euiSwitch__track"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiSwitch__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7.293 8L3.146 3.854a.5.5 0 11.708-.708L8 7.293l4.146-4.147a.5.5 0 01.708.708L8.707 8l4.147 4.146a.5.5 0 01-.708.708L8 8.707l-4.146 4.147a.5.5 0 01-.708-.708L7.293 8z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiSwitch__icon euiSwitch__icon--checked"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6.5 12a.502.502 0 01-.354-.146l-4-4a.502.502 0 01.708-.708L6.5 10.793l6.646-6.647a.502.502 0 01.708.708l-7 7A.502.502 0 016.5 12"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </span>
        </button>
        <span
          class="euiSwitch__label"
          id="random_html_id"
        >
          Show Y-axis label
        </span>
      </div>
    </div>
  </div>
</div>
`;
